### PR TITLE
Jp1/attested errors

### PR DESCRIPTION
**Pull request by darracott**
Addresses https://github.com/microsoft/scitt-ccf-ledger/issues/105

### CODE REVIEW

File: test/test_resolution.py [hunk 1]
```diff
@@ -322,6 +322,26 @@ def test_dont_cache_error(
         receipt = client.submit_claim(claim).receipt
         verify_receipt(claim, trust_store, receipt)
 
+    def test_fetch_error(
+        self,
+        client: Client,
+        did_web: DIDWebServer,
+    ):
+        """
+        Test that the ledger does not cache failed resolutions.
```
Comment (darracott) [inline test/test_resolution.py:331]: Need to update this description

File: app/fetch-did-web-doc.py [hunk 2]
```diff
@@ -26,8 +27,11 @@ def fetch_unattested(url, nonce):
         {
             "url": url,
             "nonce": nonce,
-            "status_code": response["status_code"],
-            "body": base64.b64encode(response["body"]).decode("utf-8"),
+            "result": {
+                "status": response["status_code"],
+                "body": base64.b64encode(response["body"]).decode("utf-8"),
+                "cert_chain": [],
+            },
```
Comment (plietar) [inline app/fetch-did-web-doc.py:34]: This is going to break the unattested parser in the ledger, and also doesn't work with refetchable. Given that `request` already does retries on http errors, there's no need to try and unify the two code paths together at this point. We can do that when we have the "virtual" attested-fetch, and then we'll be able to simplify this file.

File: app/fetch-did-web-doc.py [hunk 3]
```diff
@@ -120,10 +143,13 @@ def request(url, data=None, headers=None):
...
-        result = fetch_unattested(url, nonce)
-    else:
-        result = fetch_attested(url, nonce)
+    fetch = fetch_unattested if unattested else fetch_attested
+    retry = True
+    tries = 0
+    while retry and tries < FETCH_RETRIES:
+        result = fetch(url, nonce)
+        retry = refetchable(result)
+        tries += 1
```
Comment (plietar) [inline app/fetch-did-web-doc.py:152]: `fetch_unattested` uses `request`, which already has its own retry loop. IMO the retry loop here should be pushed into the fetch_attested function.

File: app/fetch-did-web-doc.py (hunk 2 above)
Comment (darracott) [reply app/fetch-did-web-doc.py:38]: Reverted the unattested changes

File: app/fetch-did-web-doc.py (hunk 3 above)
Comment (darracott) [reply app/fetch-did-web-doc.py]: I've updated this to only retry the attested fetch - I do think it's more clear to keep this level of retry here given we are inspecting the output of attested fetch to decide whether to repeat it, i.e. attested-fetch is completing and we are then repeating it. 
The attested-fetch function already has retry logic to deal with network issues, so potentially where attested-fetch is attesting to a curl error we could skip the higher level retries.

File: test/test_resolution.py [hunk 4]
```diff
@@ -318,6 +318,23 @@ def test_dont_cache_error(
         receipt = client.submit_claim(claim).receipt
         verify_receipt(claim, trust_store, receipt)
 
+    def test_fetch_error(
```
Comment (plietar) [inline test/test_resolution.py:321]: You can add
```python
@pytest.mark.skipif(os.getenv("PLATFORM") == "virtual")
```
to disable this when using unattested fetch.

File: app/fetch-did-web-doc.py (hunk 3 above)
Comment (letmaik) [reply app/fetch-did-web-doc.py]: > The attested-fetch function already has retry logic to deal with network issues, so potentially where attested-fetch is attesting to a curl error we could skip the higher level retries.

I think that's not accurate as network issues also lead to an attested fetch result (where `error` is present). Only when the enclave crashes for some reason would the existing retry logic in `fetch_attested` kick in now. I agree with @plietar that the retry loop should be inside `fetch_attested` but we can also do that in a follow-up PR.

**Review by letmaik — APPROVED** [review]

**Pull request merged**